web/web.go

```diff
@@ -0,0 +1,993 @@
+			srcTable := sessionState.conv.ToSource[t].Name
+			srcCol := sessionState.conv.ToSource[t].Cols[col]
+			srcColDef := sessionState.conv.SrcSchema[srcTable].ColDefs[srcCol]
+			// If the srcCol's type is in the map, then recalculate the Spanner type
+			// for this column using the map. Otherwise, leave the ColDef for this
+			// column as is. Note that per-column type overrides could be lost in
+			// this process -- the mapping in typeMap always takes precendence.
+			if _, found := typeMap[srcColDef.Type.Name]; found {
+				updateType(typeMap[srcColDef.Type.Name], t, col, srcTable, w)
+			}
+		}
+	}
+	updateSessionFile()
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(sessionState.conv)
+}
+
+// Actions to be performed on a column.
+// (1) Removed: true/false
+// (2) Rename: New name or empty string
+// (3) PK: "ADDED", "REMOVED" or ""
+// (4) NotNull: "ADDED", "REMOVED" or ""
+// (5) ToType: New type or empty string
+type updateCol struct {
+	Removed bool   `json:"Removed"`
+	Rename  string `json:"Rename"`
+	PK      string `json:"PK"`
+	NotNull string `json:"NotNull"`
+	ToType  string `json:"ToType"`
+}
+
+type updateTable struct {
+	UpdateCols map[string]updateCol `json:"UpdateCols"`
+}
+
+// updateTableSchema updates the Spanner schema.
+// Following actions can be performed on a specified table:
+// (1) Remove column
+// (2) Rename column
+// (3) Add or Remove Primary Key
+// (4) Add or Remove NotNull constraint
+// (5) Update Spanner type
+func updateTableSchema(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	var t updateTable
+	table := r.FormValue("table")
+	err = json.Unmarshal(reqBody, &t)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+	srcTableName := sessionState.conv.ToSource[table].Name
+	for colName, v := range t.UpdateCols {
+		if v.Removed {
+			//TODO:(searce) and handling rollback uniformly with return rollback(err)
+			err, status := canRemoveColumn(colName, table)
+			if err != nil {
+				err = rollback(err)
+				http.Error(w, fmt.Sprintf("%v", err), status)
+				return
+			}
+			removeColumn(table, colName, srcTableName)
+			continue
+		}
+		if v.Rename != "" && v.Rename != colName {
+			err, status := canRenameOrChangeType(colName, table)
+			if err != nil {
+				err = rollback(err)
+				http.Error(w, fmt.Sprintf("%v", err), status)
+				return
+			}
+			renameColumn(v.Rename, table, colName, srcTableName)
+			colName = v.Rename
+		}
+		if v.PK != "" {
+			http.Error(w, "HarbourBridge currently doesn't support editing primary keys", http.StatusNotImplemented)
+			return
+		}
+
+		typeChange, err := isTypeChanged(v.ToType, table, colName, srcTableName)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
```

Comment:
nit; I wonder why not do this inside "if v.ToType != "" {}"

Reply:
changed accordingly